url5a01f2c1d9_core9_core_tgz-pJ8-io_core9_core_boot_BootstrapFrameworkJ8/src/net/xeoh/nexus/DefaultNexus.java

```diff
@@ -0,0 +1,304 @@
+ * above copyright notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ * 
+ * Neither the name of the author nor the names of its contributors may be used to endorse or
+ * promote products derived from this software without specific prior written permission.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * 
+ */
+package net.xeoh.nexus;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import net.xeoh.nexus.options.Option;
+import net.xeoh.nexus.states.State;
+import net.xeoh.nexus.states.StateDisabled;
+
+/**
+ * A default implementation of the {@link Nexus} interface. Unless you have specific needs,
+ * use this class.
+ * 
+ * @author Ralf Biedert
+ * @since 1.0
+ */
+public class DefaultNexus implements Nexus {
+    /** All services we know */
+    final ConcurrentLinkedQueue<Service> services = new ConcurrentLinkedQueue<Service>();
+
+    /** All service listeners we know */
+    final ConcurrentLinkedQueue<ServiceListener> serviceListeners = new ConcurrentLinkedQueue<ServiceListener>();
+    
+    /** A quick look cache for commons requests */
+    final Map<Class<?>, ConcurrentLinkedQueue<Service>> cache = new ConcurrentHashMap<Class<?>, ConcurrentLinkedQueue<Service>>();
+
+    
+    /**
+     * Returns the appropriate queue for the given class.
+     * 
+     * @since 1.0
+     * @param clazz The class to get the queue for.
+     * @return The queue holding the elements.
+     */
+    private ConcurrentLinkedQueue<Service> queueFor(Class<?> clazz) {
+        synchronized (this.cache) {
+            // Check if we already have a queue
+            final ConcurrentLinkedQueue<Service> queue = this.cache.get(clazz);
+            if(queue != null) return queue;
+            
+            // If not, create it and place it into the cache
+            final ConcurrentLinkedQueue<Service> rval = new ConcurrentLinkedQueue<Service>();
+            this.cache.put(clazz, rval);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **70 and 75**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **71**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).